src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestHeaders.cs

```diff
@@ -24,33 +68,42 @@ private static long ParseContentLength(string value)
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private void SetValueUnknown(string key, in StringValues value)
+        private void AppendContentLength(Span<byte> value)
         {
-            Unknown[key] = value;
-        }
+            if (_contentLength.HasValue)
+            {
+                BadHttpRequestException.Throw(RequestRejectionReason.MultipleContentLengths);
+            }
 
-        public unsafe void Append(Span<byte> name, string value)
-        {
-            fixed (byte* namePtr = name)
+            if (!HeaderUtilities.TryParseNonNegativeInt64(value, out var parsed))
             {
-                Append(namePtr, name.Length, value);
+                BadHttpRequestException.Throw(RequestRejectionReason.InvalidContentLength, value.GetAsciiOrUTF8StringNonNullCharacters());
             }
+
+            _contentLength = parsed;
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void SetValueUnknown(string key, in StringValues value)
+        {
+            Unknown[key] = value;
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
-        private unsafe void AppendUnknownHeaders(byte* pKeyBytes, int keyLength, string value)
+        private unsafe void AppendUnknownHeaders(Span<byte> name, string valueString)
         {
-            string key = new string('\0', keyLength);
+            string key = new string('\0', name.Length);
```

Comment:
(I know this is existing code. It doesn't need to be addressed as part of this PR, but it should be addressed at some point.)

Reply:
Is a messy area https://github.com/aspnet/AspNetCore/pull/8374#discussion_r264014541 and should be picked up in a follow up PR as its just what it does currently. Equally the headers should be passed as `ReadOnlySpan<byte>` rather than `Span<byte>` and certainly before this interface is made public public rather than pubternal
```csharp
public partial interface IHttpHeadersHandler
{
    void OnHeader(System.Span<byte> name, System.Span<byte> value);
    void OnHeadersComplete();
}
```

Reply:
https://github.com/aspnet/AspNetCore/issues/8978 for string

https://github.com/aspnet/AspNetCore/issues/8979 for interfaces